src/worker.rs

```diff
@@ -149,37 +207,34 @@ impl Worker {
                 }) => match serde_json::to_value(req.req.clone()) {
                     Ok(json) => {
+                        let policy_name = policy_evaluator.policy.id.clone();
+                        let policy_mode = policy_mode.clone();
                         let start_time = Instant::now();
-                        let resp = policy_evaluator.validate(ValidateRequest::new(json));
+                        let allowed_to_mutate = *allowed_to_mutate;
+                        let validation_response =
+                            policy_evaluator.validate(ValidateRequest::new(json));
                         let policy_evaluation_duration = start_time.elapsed();
-                        let accepted = resp.allowed;
-                        let mutated = resp.patch.is_some();
-                        let error_code = if let Some(status) = &resp.status {
+                        let error_code = if let Some(status) = &validation_response.status {
                             status.code
                         } else {
                             None
                         };
-                        let resp = if mutated && allowed_to_mutate.as_ref() == Some(&false) {
-                            ValidationResponse {
-                                allowed: false,
-                                status: Some(ValidationResponseStatus {
-                                    message: Some(format!("Request rejected by policy {}. The policy attempted to mutate the request, but it is currently configured to not allow mutations.", &req.policy_id)),
-                                    code: None,
-                                }),
-                                // if `allowed_to_mutate` is false, we are in a validating webhook. If we send a patch, k8s will fail because validating webhook do not expect this field
-                                patch: None,
-                                patch_type: None,
-                                ..resp
-                            }
-                        } else {
-                            resp
-                        };
-                        let res = req.resp_chan.send(Some(resp));
+                        let accepted = validation_response.allowed;
+                        let mutated = validation_response.patch.is_some();
+                        let validation_response = Worker::validation_response_with_constraints(
+                            &req.policy_id,
+                            &policy_mode,
+                            allowed_to_mutate,
+                            validation_response,
+                        );
+                        let res = req.resp_chan.send(Some(validation_response));
                         let policy_evaluation = metrics::PolicyEvaluation {
-                            policy_name: policy_evaluator.policy.id.clone(),
+                            policy_name,
+                            policy_mode: policy_mode.into(),
```

Comment:
I think we have to do more changes. Right now the metrics are unable to distinguish between protecting policies accepting requests, monitoring policies that would one day accept requests and - most important of all - monitoring policies that would one day reject a request.

I think that when dealing with monitoring policies, we should retain the "future status when protect mode is enabled", and expose that inside of the metrics.
To simplify, we should emit a metric event about a rejection. The reason, IMHO, is to allow an operator to quickly identify the monitoring policies that, once switched to protect mode, would instead generate a flood of rejections.

Right now, this information is buried only inside of jaeger traces

Reply:
Being this part of the baggage of the metric, the user can use this `policy_mode` to filter the evaluation metric done in different modes, and what they would have answered in the case of `policy_mode: monitor`.

Reply:
The metric is exposing the value of `accepted`, which is set on line 226:

```rust
let accepted = validation_response.allowed;
```

The `validation_response` object is returned by the function `validation_response_with_constraints`.

Inside of `validation_response_with_constraints`, a policy that has `monitor` mode will always return a response with `allowed` set to `true`. The information about the possible outcome when the policy is switched to "protect" mode is lost.

Reply:
Hm, no, it's not lost. Line 226 goes all the way to 242 (the metric baggage)... Am I missing something?

Reply:
Just to give more information as well: `validation_response_with_constraints` does not mutate its input. It builds a new `ValidationResponse` object, that copies some arguments inside its structure attributes, but is a new object.

Reply:
No, it's my fault. Looking at the whole file I see what is happening. I got mislead by the github diff view + the overwriting of the `validation_response`.

It's all good... sorry about the noise

Reply:
Thanks 🙏